src/downloader/worker.rs

```diff
@@ -0,0 +1,138 @@
+                _ = tokio::time::sleep(delay) => {},
+                _ = req.cancel.cancelled() => {
+                    req.status.send(Status::Failed).ok();
+                    req.result.send(Err(Error::Download(DownloadError::Cancelled))).ok();
+                    return;
+                }
+            }
+        }
+
+        match download(client.clone(), &mut req).await {
+            Ok(file) => {
+                req.status.send(Status::Completed).ok();
+                req.result.send(Ok(file)).ok();
+                return;
+            }
+            Err(e) => {
+                if should_retry(&e) {
+                    last_error = Some(e);
+                    continue;
+                }
+
+                let status = if matches!(e, Error::Download(DownloadError::Cancelled)) {
+                    Status::Cancelled
+                } else {
+                    Status::Failed
+                };
+                req.status.send(status).ok();
+                req.result.send(Err(e)).ok();
+                return;
+            }
+        }
+    }
+}
+
+async fn download(client: Client, req: &mut DownloadRequest) -> Result<File, Error> {
+    let update_progress = |bytes_downloaded: u64, total_bytes: Option<u64>| {
+        req.status
+            .send(Status::InProgress(DownloadProgress {
+                bytes_downloaded,
+                total_bytes,
+            }))
+            .ok();
+    };
+
+    let mut response = client
+        .get(req.url.as_ref())
+        .send()
+        .await?
+        .error_for_status()?;
+    let total_bytes = response.content_length();
+    let mut bytes_downloaded = 0u64;
+
+    // Create the destination directory if it doesn't exist
+    if let Some(parent) = req.destination.parent() {
+        tokio::fs::create_dir_all(parent).await?;
+    }
+    let mut file = File::create(&req.destination).await?;
+
+    update_progress(bytes_downloaded, total_bytes);
+    loop {
+        tokio::select! {
+            _ = req.cancel.cancelled() => {
+                drop(file); // Manually drop the file handle to ensure that deletion doesn't fail
+                tokio::fs::remove_file(&req.destination).await?;
+                return Err(Error::Download(DownloadError::Cancelled));
+            }
+            chunk = response.chunk() => {
+                match chunk {
+                    Ok(Some(chunk)) => {
+                        file.write_all(&chunk).await?;
+                        bytes_downloaded += chunk.len() as u64;
+                        update_progress(bytes_downloaded, total_bytes);
+                    }
+                    Ok(None) => break,
+                    Err(e) => {
+                        drop(file); // Manually drop the file handle to ensure that deletion doesn't fail
+                        tokio::fs::remove_file(&req.destination).await?;
+                        return Err(Error::Reqwest(e))
+                    },
+                }
+            }
+        }
+    }
```

Comment:
The manual `drop`, `remove_file`, `req.status.send`, `req.result.send` are error prone, do we have any ideas how to automatically do these